Arch.LowLevel/UnsafeQueue.cs

```diff
@@ -119,9 +124,29 @@ public void TrimExcess()
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
     public void EnsureCapacity(int newCapacity)
     {
-        if (newCapacity <= _capacity)
+        EnsureCapacity(newCapacity, false);
+    }
+
+    /// <summary>
+    ///     Ensures the capacity of this instance. 
+    /// </summary>
+    /// <param name="newCapacity">The new capacity.</param>
+    /// <param name="allowShrinking">Indicates if the new capacity may be smaller than current capacity</param>
+    /// <exception cref="ArgumentException"></exception>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void EnsureCapacity(int newCapacity, bool allowShrinking)
+    {
+        if (allowShrinking)
         {
-            return;
+            if (newCapacity < _count)
+                throw new ArgumentOutOfRangeException(nameof(newCapacity), "newCapacity cannot be smaller than _count");
+        }
+        else
+        {
+            if (newCapacity <= _capacity)
+            {
+                return;
+            }
```

Comment:
Whole PR looks good so far :) 
But is this necessary? I think it would be easier to understand if we seperate the logic. Probably we should remove that overload ^^

Reply:
I didn't want to duplicate all of the code further down this method into both the `TrimExcess` and `EnsureCapacity` methods.

I've just pushed up an alternative - renaming the `EnsureCapacity` overload to `SetCapacity` which I think more clearly states what it's for. What do you think of that?